File: sift/notes.py
from __future__ import annotations

import argparse
import hashlib
import json
import shutil
import subprocess
import sys
import tempfile
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path


PROJECT_ROOT = Path(__file__).resolve().parent
DEFAULT_EXPORT_DIR = PROJECT_ROOT / "docs" / "apple-notes"
DEFAULT_DB_DIR = PROJECT_ROOT / "db"


_NOTES_EXPORT_JXA = r"""
const Notes = Application("Notes");
const exported = [];
let skipped = 0;

for (const note of Notes.notes()) {
    try {
        exported.push({
            id: String(note.id()),
            title: String(note.name() || ""),
            body: String(note.plaintext() || "")
        });
    } catch (error) {
        skipped += 1;
    }
}

JSON.stringify({notes: exported, skipped});
""".strip()


class NotesExportError(RuntimeError):
    pass


@dataclass(frozen=True)
class AppleNote:
    note_id: str
    title: str
    body: str


FetchNotes = Callable[[], tuple[list[AppleNote], int]]
RebuildIndex = Callable[[Path, Path], None]


def parse_export_payload(payload: str) -> tuple[list[AppleNote], int]:
    try:
        data = json.loads(payload)
        if not isinstance(data, dict):
            raise TypeError

        raw_notes = data["notes"]
        skipped = data["skipped"]
        if (
            not isinstance(raw_notes, list)
            or not isinstance(skipped, int)
            or isinstance(skipped, bool)
            or skipped < 0
        ):
            raise TypeError

        notes = []
        note_ids = set()
        for item in raw_notes:
            if not isinstance(item, dict) or set(item) != {"id", "title", "body"}:
                raise TypeError
            note_id = item["id"]
            title = item["title"]
            body = item["body"]
            if not all(isinstance(value, str) for value in (note_id, title, body)):
                raise TypeError
            if not note_id or note_id in note_ids:
                raise TypeError
            note_ids.add(note_id)
            notes.append(AppleNote(note_id, title, body))
        return notes, skipped
    except (KeyError, TypeError, ValueError) as error:
        raise NotesExportError("Notes returned invalid export data.") from error


def fetch_notes() -> tuple[list[AppleNote], int]:
    if sys.platform != "darwin":
        raise NotesExportError("Apple Notes sync is macOS-only.")

    try:
        result = subprocess.run(
            ["osascript", "-l", "JavaScript", "-e", _NOTES_EXPORT_JXA],
            check=True,
            capture_output=True,
            text=True,
        )
    except FileNotFoundError as error:
        raise NotesExportError(
            "Apple Notes sync failed because osascript is unavailable."
        ) from error
    except subprocess.CalledProcessError as error:
        stderr = error.stderr if isinstance(error.stderr, str) else ""
        authorization_signals = ("-1743", "not authorized", "not permitted")
        if any(signal in stderr.lower() for signal in authorization_signals):
            raise NotesExportError(
                "Apple Notes sync needs macOS Automation permission. Allow "
                "access under System Settings > Privacy & Security > "
                "Automation."
            ) from error
        raise NotesExportError("Apple Notes automation failed.") from error

    return parse_export_payload(result.stdout)


def note_filename(note_id: str) -> str:
    digest = hashlib.sha256(note_id.encode("utf-8")).hexdigest()[:20]
    return f"{digest}.txt"


def write_export(notes: list[AppleNote], export_dir: Path) -> None:
    export_dir.parent.mkdir(parents=True, exist_ok=True)
    staging = Path(tempfile.mkdtemp(prefix=".apple-notes-", dir=export_dir.parent))
    backup = None
    try:
        for note in notes:
            content = f"Title: {note.title.strip()}\n\n{note.body.strip()}\n"
            (staging / note_filename(note.note_id)).write_text(
                content,
                encoding="utf-8",
            )
        if export_dir.exists():
            backup = Path(
                tempfile.mkdtemp(
                    prefix=".apple-notes-backup-",
                    dir=export_dir.parent,
                )
            )
            backup.rmdir()
            export_dir.replace(backup)
        try:
            staging.replace(export_dir)
        except Exception:
            if backup is not None:
                backup.replace(export_dir)
            raise
        if backup is not None:
            shutil.rmtree(backup)
    except Exception:
        shutil.rmtree(staging, ignore_errors=True)
        raise


def rebuild_index(
    docs_dir: Path,
    db_dir: Path,
    embed_model: str = "nomic-embed-text",
) -> None:
    import main as rag

    rag.get_vectorstore(docs_dir, db_dir, embed_model)


def _unused_hidden_path(parent: Path, prefix: str) -> Path:
    parent.mkdir(parents=True, exist_ok=True)
    path = Path(tempfile.mkdtemp(prefix=prefix, dir=parent))
    path.rmdir()
    return path


def _remove_path(path: Path) -> None:
    if path.is_dir() and not path.is_symlink():
        shutil.rmtree(path, ignore_errors=True)
    elif path.exists() or path.is_symlink():
        try:
            path.unlink()
        except OSError:
            pass


def _refresh_index(
    docs_dir: Path,
    db_dir: Path,
    rebuild: RebuildIndex,
) -> None:
    staging = _unused_hidden_path(
        db_dir.parent,
        f".{db_dir.name}-staging-",
    )
    backup: Path | None = None

    try:
        rebuild(docs_dir, staging)
        if not staging.exists():
            raise RuntimeError("Index rebuild did not create a database.")

        if db_dir.exists() or db_dir.is_symlink():
            backup = _unused_hidden_path(
                db_dir.parent,
                f".{db_dir.name}-backup-",
            )
            db_dir.replace(backup)

        try:
            staging.replace(db_dir)
        except Exception:
            if backup is not None:
                backup.replace(db_dir)
            raise

        if backup is not None:
            _remove_path(backup)
    except Exception:
        _remove_path(staging)
        if (
            backup is not None
            and (backup.exists() or backup.is_symlink())
            and not (db_dir.exists() or db_dir.is_symlink())
        ):
            backup.replace(db_dir)
        if backup is not None:
            _remove_path(backup)
        raise


def sync_notes(
    export_dir: Path = DEFAULT_EXPORT_DIR,
    db_dir: Path = DEFAULT_DB_DIR,
    *,
    fetch: FetchNotes | None = None,
    rebuild: RebuildIndex | None = None,
) -> tuple[int, int]:
    fetch_fn = fetch_notes if fetch is None else fetch
    rebuild_fn = rebuild_index if rebuild is None else rebuild

    notes, skipped = fetch_fn()
    try:
        write_export(notes, export_dir)
    except NotesExportError:
        raise
    except Exception as error:
        raise NotesExportError("Apple Notes export failed.") from error

    try:
        _refresh_index(export_dir.parent, db_dir, rebuild_fn)
    except Exception as error:
        raise NotesExportError(
            "Notes were exported, but the index rebuild failed."
        ) from error

    return len(notes), skipped


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(description="Sync Apple Notes into local RAG.")
    subparsers = parser.add_subparsers(dest="command", required=True)
    subparsers.add_parser("sync", help="Export notes and rebuild the index.")
    args = parser.parse_args(argv)

    if args.command == "sync":
        try:
            exported, skipped = sync_notes()
        except NotesExportError as error:
            print(f"Error: {error}", file=sys.stderr)
            return 1
        print(
            f"Exported {exported} notes; skipped {skipped}; index rebuilt."
        )
        return 0

    parser.error("a command is required")


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/notes/notes.go
package notes

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const ExportJXA = `
const Notes = Application("Notes");
const exported = [];
let skipped = 0;

for (const note of Notes.notes()) {
    try {
        exported.push({
            id: String(note.id()),
            title: String(note.name() || ""),
            body: String(note.plaintext() || "")
        });
    } catch (error) {
        skipped += 1;
    }
}

JSON.stringify({notes: exported, skipped});
`

type AppleNote struct {
	ID    string
	Title string
	Body  string
}

type ExportError struct {
	Msg string
}

func (e *ExportError) Error() string {
	return e.Msg
}

type FetchFunc func() ([]AppleNote, int, error)
type RebuildFunc func(docsDir, dbDir string) error
type runFunc func(name string, args []string) (stdout, stderr string, err error)

var (
	writeFile   = os.WriteFile
	renamePath  = os.Rename
	writeExport = WriteExport
)

func NoteFilename(noteID string) string {
	sum := sha256.Sum256([]byte(noteID))
	return hex.EncodeToString(sum[:])[:20] + ".txt"
}

func ParseExportPayload(payload string) ([]AppleNote, int, error) {
	invalid := &ExportError{Msg: "Notes returned invalid export data."}

	var root map[string]json.RawMessage
	if err := json.Unmarshal([]byte(payload), &root); err != nil {
		return nil, 0, invalid
	}

	rawNotes, ok := root["notes"]
	if !ok {
		return nil, 0, invalid
	}
	rawSkipped, ok := root["skipped"]
	if !ok {
		return nil, 0, invalid
	}

	skipped, err := parseSkipped(rawSkipped)
	if err != nil {
		return nil, 0, invalid
	}

	var items []json.RawMessage
	if err := json.Unmarshal(rawNotes, &items); err != nil {
		return nil, 0, invalid
	}

	notes := make([]AppleNote, 0, len(items))
	seen := make(map[string]struct{})
	for _, item := range items {
		var fields map[string]json.RawMessage
		if err := json.Unmarshal(item, &fields); err != nil {
			return nil, 0, invalid
		}
		if len(fields) != 3 {
			return nil, 0, invalid
		}
		id, err := parseJSONString(fields["id"])
		if err != nil {
			return nil, 0, invalid
		}
		title, err := parseJSONString(fields["title"])
		if err != nil {
			return nil, 0, invalid
		}
		body, err := parseJSONString(fields["body"])
		if err != nil {
			return nil, 0, invalid
		}
		if id == "" {
			return nil, 0, invalid
		}
		if _, dup := seen[id]; dup {
			return nil, 0, invalid
		}
		seen[id] = struct{}{}
		notes = append(notes, AppleNote{ID: id, Title: title, Body: body})
	}
	return notes, skipped, nil
}

func parseJSONString(raw json.RawMessage) (string, error) {
	if len(raw) == 0 || strings.TrimSpace(string(raw)) == "null" {
		return "", errors.New("missing")
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", err
	}
	return value, nil
}

func parseSkipped(raw json.RawMessage) (int, error) {
	trimmed := strings.TrimSpace(string(raw))
	if trimmed == "true" || trimmed == "false" {
		return 0, errors.New("bool")
	}
	var skipped int
	if err := json.Unmarshal(raw, &skipped); err != nil {
		return 0, err
	}
	if skipped < 0 {
		return 0, errors.New("negative")
	}
	return skipped, nil
}

func FetchNotes(platform string, run runFunc) ([]AppleNote, int, error) {
	if platform != "darwin" {
		return nil, 0, &ExportError{Msg: "Apple Notes sync is macOS-only."}
	}

	stdout, stderr, err := run("osascript", []string{"-l", "JavaScript", "-e", strings.TrimSpace(ExportJXA)})
	if err != nil {
		if errors.Is(err, exec.ErrNotFound) {
			return nil, 0, &ExportError{Msg: "Apple Notes sync failed because osascript is unavailable."}
		}
		lower := strings.ToLower(stderr)
		for _, signal := range []string{"-1743", "not authorized", "not permitted"} {
			if strings.Contains(lower, signal) {
				return nil, 0, &ExportError{Msg: "Apple Notes sync needs macOS Automation permission. Allow access under System Settings > Privacy & Security > Automation."}
			}
		}
		return nil, 0, &ExportError{Msg: "Apple Notes automation failed."}
	}
	return ParseExportPayload(stdout)
}

func WriteExport(notes []AppleNote, exportDir string) error {
	parent := filepath.Dir(exportDir)
	if err := os.MkdirAll(parent, 0o755); err != nil {
		return err
	}
	staging, err := os.MkdirTemp(parent, ".apple-notes-")
	if err != nil {
		return err
	}

	backup := ""
	cleanupStaging := true
	defer func() {
		if cleanupStaging {
			os.RemoveAll(staging)
		}
	}()

	for _, note := range notes {
		content := "Title: " + strings.TrimSpace(note.Title) + "\n\n" + strings.TrimSpace(note.Body) + "\n"
		if err := writeFile(filepath.Join(staging, NoteFilename(note.ID)), []byte(content), 0o644); err != nil {
			return err
		}
	}

	if exists(exportDir) {
		backup, err = unusedHiddenPath(parent, ".apple-notes-backup-")
		if err != nil {
			return err
		}
		if err := renamePath(exportDir, backup); err != nil {
			return err
		}
	}

	if err := renamePath(staging, exportDir); err != nil {
		if backup != "" {
			_ = renamePath(backup, exportDir)
		}
		return err
	}
	cleanupStaging = false
	if backup != "" {
		os.RemoveAll(backup)
	}
	return nil
}

func Sync(exportDir, dbDir string, fetch FetchFunc, rebuild RebuildFunc) (int, int, error) {
	notes, skipped, err := fetch()
	if err != nil {
		return 0, 0, err
	}

	if err := writeExport(notes, exportDir); err != nil {
		var exportErr *ExportError
		if errors.As(err, &exportErr) {
			return 0, 0, err
		}
		return 0, 0, &ExportError{Msg: "Apple Notes export failed."}
	}

	if err := refreshIndex(filepath.Dir(exportDir), dbDir, rebuild); err != nil {
		return 0, 0, &ExportError{Msg: "Notes were exported, but the index rebuild failed."}
	}
	return len(notes), skipped, nil
}

func refreshIndex(docsDir, dbDir string, rebuild RebuildFunc) error {
	staging, err := unusedHiddenPath(filepath.Dir(dbDir), "."+filepath.Base(dbDir)+"-staging-")
	if err != nil {
		return err
	}

	backup := ""
	defer func() {
		os.RemoveAll(staging)
		if backup != "" && exists(backup) && !exists(dbDir) {
			_ = renamePath(backup, dbDir)
		}
		if backup != "" {
			removePath(backup)
		}
	}()

	if err := rebuild(docsDir, staging); err != nil {
		return err
	}
	if !exists(staging) {
		return errors.New("Index rebuild did not create a database.")
	}

	if exists(dbDir) {
		backup, err = unusedHiddenPath(filepath.Dir(dbDir), "."+filepath.Base(dbDir)+"-backup-")
		if err != nil {
			return err
		}
		if err := renamePath(dbDir, backup); err != nil {
			return err
		}
	}

	if err := renamePath(staging, dbDir); err != nil {
		if backup != "" {
			_ = renamePath(backup, dbDir)
		}
		return err
	}
	if backup != "" {
		removePath(backup)
		backup = ""
	}
	staging = ""
	return nil
}

func unusedHiddenPath(parent, prefix string) (string, error) {
	if err := os.MkdirAll(parent, 0o755); err != nil {
		return "", err
	}
	path, err := os.MkdirTemp(parent, prefix)
	if err != nil {
		return "", err
	}
	if err := os.Remove(path); err != nil {
		return "", err
	}
	return path, nil
}

func exists(path string) bool {
	_, err := os.Lstat(path)
	return err == nil
}

func removePath(path string) {
	info, err := os.Lstat(path)
	if err != nil {
		return
	}
	if info.IsDir() && info.Mode()&os.ModeSymlink == 0 {
		os.RemoveAll(path)
		return
	}
	os.Remove(path)
}
